openhands/sdk/conversation/conversation.py

```diff
@@ -108,8 +120,54 @@ def run(self) -> None:
             # and check will we be able to execute .send_message
             # BEFORE the .run loop finishes?
             with self.state:
+                # clear the flag before calling agent.step() (user approved)
+                if self.state.waiting_for_confirmation:
+                    self.state.waiting_for_confirmation = False
+
                 # step must mutate the SAME state object
                 self.agent.step(self.state, on_event=self._on_event)
+
+            # In confirmation mode, stop after one iteration if waiting for confirmation
+            if self.state.waiting_for_confirmation:
+                break
+
             iteration += 1
             if iteration >= self.max_iteration_per_run:
                 break
+
+    def get_pending_actions(self) -> list[ActionEvent]:
+        """Get all actions that are waiting for confirmation from the agent."""
+        return self.agent.get_pending_actions(self.state)
+
+    def set_confirmation_mode(self, enabled: bool) -> None:
+        """Enable or disable confirmation mode and store it in conversation state."""
+        with self.state:
+            self.state.confirmation_mode = enabled
+        logger.info(f"Confirmation mode {'enabled' if enabled else 'disabled'}")
+
+    def reject_pending_actions(self, reason: str = "User rejected the action") -> None:
+        """Reject all pending actions from the agent.
+
+        This is a non-invasive method to reject actions between run() calls.
+        Also clears the waiting_for_confirmation flag.
+        """
+        pending_actions = self.agent.get_pending_actions(self.state)
+
+        with self.state:
+            # Always clear the waiting_for_confirmation flag
+            self.state.waiting_for_confirmation = False
+
+            if not pending_actions:
+                logger.warning("No pending actions to reject")
+                return
+
+            for action_event in pending_actions:
+                # Create rejection observation
+                rejection_event = UserRejectsObservation(
+                    action_id=action_event.id,
+                    tool_name=action_event.tool_name,
+                    tool_call_id=action_event.tool_call_id,
+                    rejection_reason=reason,
+                )
+                self._on_event(rejection_event)
+                logger.info(f"Rejected pending action: {action_event} - {reason}")
```

Comment:
Ideally we should have the ability to confirm each action, rrather than batch accept/reject?

Reply:
I actually wonder if that's too much overhead for the user. Its also possible commands are tough to decouple. 

In the CLI the user can reject the action with a reason, so the agent could retry by removing or modifying a subset of actions?